src/transfers.rs

```diff
@@ -30,48 +36,146 @@ struct Rule {
     to_stop_id: String,
     transfer_time: Option<u32>,
 }
+/// Represents the type of transfers to generate
+#[derive(PartialEq, Debug)]
+pub enum TransfersMode {
+    /// `All` will generate all transfers
+    All,
+    /// `IntraContributor` will generate transfers between stop points belonging to the
+    /// same contributor
+    IntraContributor,
+    /// `InterContributor` will generate transfers between stop points belonging to
+    /// differents contributors only
+    InterContributor,
+}
 
 type TransferMap = HashMap<(Idx<StopPoint>, Idx<StopPoint>), Transfer>;
 
+fn stop_points_need_transfer(
+    model: &Model,
+    from_idx: Idx<StopPoint>,
+    to_idx: Idx<StopPoint>,
+    transfers_mode: &TransfersMode,
+    report_opt: Option<&mut Report>,
+) -> bool {
+    if *transfers_mode == TransfersMode::All {
+        return true;
+    }
+    let from_contributor: BTreeSet<Idx<Contributor>> = model.get_corresponding_from_idx(from_idx);
+    let to_contributor: BTreeSet<Idx<Contributor>> = model.get_corresponding_from_idx(to_idx);
+    if from_contributor.is_empty() {
+        if let Some(report) = report_opt {
+            report.add_warning(
+                format!(
+                    "stop point {} belongs to none of the trips and will not generate any transfer",
+                    model.stop_points[from_idx].id
+                ),
+                ReportType::TransferOnUnreferencedStop,
+            );
+        }
+        return false;
+    }
+    if to_contributor.is_empty() {
+        if let Some(report) = report_opt {
+            report.add_warning(
+                format!(
+                    "stop point {} belongs to none of the trips and will not generate any transfer",
+                    model.stop_points[to_idx].id
+                ),
+                ReportType::TransferOnUnreferencedStop,
+            );
+        }
+        return false;
+    }
+    let same_contributor = from_contributor == to_contributor;
+    same_contributor && *transfers_mode == TransfersMode::IntraContributor
+        || !same_contributor && *transfers_mode == TransfersMode::InterContributor
```

Comment:
```rust
match transfers_mode {
    TransfersMode::All => true,
    TransfersMode::IntraContributor => from_contributor == to_contributor,
    TransfersMode::InterContributor => from_contributor != to_contributor,
}
```

As all values are tested, if you modify TransfersMode, you will not have nasty bugs, you'll have a compile error.

Reply:
:+1: